File: src/context/parse_context.py
#!/usr/bin/env python3
"""Combine imported assessment material into one bounded Markdown context file."""
from __future__ import annotations

import argparse
import csv
import html
import json
import re
from pathlib import Path

MAX_FILE_BYTES = 8 * 1024 * 1024
MAX_TEXT_CHARS = 200_000


def extract(path: Path) -> tuple[str, str]:
    if path.stat().st_size > MAX_FILE_BYTES:
        return "", "skipped: file exceeds 8 MB"
    suffix = path.suffix.lower()
    try:
        if suffix == ".pdf":
            from pypdf import PdfReader  # optional
            return "\n\n".join((page.extract_text() or "") for page in PdfReader(str(path)).pages), "PDF"
        if suffix == ".docx":
            from docx import Document  # optional
            return "\n".join(p.text for p in Document(str(path)).paragraphs), "Word document"
        if suffix in {".xlsx", ".xlsm"}:
            from openpyxl import load_workbook  # optional
            book = load_workbook(path, read_only=True, data_only=True)
            lines = []
            for sheet in book.worksheets:
                lines.append(f"Sheet: {sheet.title}")
                lines.extend(" | ".join("" if value is None else str(value) for value in row) for row in sheet.iter_rows(values_only=True))
            return "\n".join(lines), "Excel workbook"
        if suffix == ".pptx":
            from pptx import Presentation  # optional
            deck = Presentation(str(path))
            return "\n".join(shape.text for slide in deck.slides for shape in slide.shapes if hasattr(shape, "text")), "PowerPoint presentation"
        if suffix in {".png", ".jpg", ".jpeg", ".webp", ".bmp", ".tiff"}:
            from PIL import Image  # optional
            import pytesseract  # optional
            return pytesseract.image_to_string(Image.open(path)), "image OCR"
        raw = path.read_text(encoding="utf-8", errors="replace")
        if suffix == ".json":
            raw = json.dumps(json.loads(raw), indent=2, ensure_ascii=False)
        elif suffix in {".html", ".htm"}:
            raw = re.sub(r"<(script|style)[^>]*>.*?</\1>", "", raw, flags=re.I | re.S)
            raw = html.unescape(re.sub(r"<[^>]+>", " ", raw))
        elif suffix in {".csv", ".tsv"}:
            delimiter = "\t" if suffix == ".tsv" else ","
            rows = list(csv.reader(raw.splitlines(), delimiter=delimiter))[:2000]
            raw = "\n".join(" | ".join(cell.replace("|", "\\|") for cell in row) for row in rows)
        return raw, suffix.lstrip(".").upper() or "text"
    except Exception as exc:
        return "", f"skipped: {exc}"


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--output", required=True)
    parser.add_argument("files", nargs="+")
    args = parser.parse_args()
    sections = ["# Penetration Testing Context", "", "Generated from imported source material.", ""]
    parsed = 0
    for raw_path in args.files:
        path = Path(raw_path)
        text, kind = extract(path)
        sections.extend([f"## {path.name}", "", f"_Source: `{path}` · {kind}_", ""])
        if text.strip():
            clean = text.strip()[:MAX_TEXT_CHARS]
            sections.extend(["```text", clean, "```", ""])
            parsed += 1
        else:
            sections.extend([f"> {kind}", ""])
    output = Path(args.output)
    output.parent.mkdir(parents=True, exist_ok=True)
    output.write_text("\n".join(sections), encoding="utf-8")
    print(json.dumps({"ok": True, "parsed": parsed, "total": len(args.files), "output": str(output)}))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/app/services/assessment/parse-context.js
"use strict";

/* Bounded context extraction for the workspace runtime. */

const fs = require("node:fs");
const path = require("node:path");

const MAX_FILE_BYTES = 8 * 1024 * 1024;
const MAX_TEXT_CHARS = 200_000;

function csvRows(raw, delimiter) {
  return raw.split(/\r?\n/).slice(0, 2000).map((line) => {
    const cells = [];
    let cell = "";
    let quoted = false;
    for (let index = 0; index < line.length; index += 1) {
      const char = line[index];
      if (char === '"') {
        if (quoted && line[index + 1] === '"') { cell += '"'; index += 1; }
        else quoted = !quoted;
      } else if (char === delimiter && !quoted) { cells.push(cell.replace(/\|/g, "\\|")); cell = ""; }
      else cell += char;
    }
    cells.push(cell.replace(/\|/g, "\\|"));
    return cells.join(" | ");
  });
}

function extract(filePath) {
  const stats = fs.statSync(filePath);
  if (stats.size > MAX_FILE_BYTES) return { text: "", kind: "skipped: file exceeds 8 MB" };
  const suffix = path.extname(filePath).toLowerCase();
  try {
    let raw = fs.readFileSync(filePath, "utf8");
    if (suffix === ".json") raw = JSON.stringify(JSON.parse(raw), null, 2);
    else if ([".html", ".htm"].includes(suffix)) raw = raw.replace(/<(script|style)[^>]*>[\s\S]*?<\/\1>/gi, "").replace(/<[^>]+>/g, " ").replace(/&lt;/g, "<").replace(/&gt;/g, ">").replace(/&amp;/g, "&").replace(/&quot;/g, '"').replace(/&#39;/g, "'");
    else if (suffix === ".csv") raw = csvRows(raw, ",").join("\n");
    else if (suffix === ".tsv") raw = csvRows(raw, "\t").join("\n");
    else if ([".pdf", ".docx", ".xlsx", ".xlsm", ".pptx", ".png", ".jpg", ".jpeg", ".webp", ".bmp", ".tiff"].includes(suffix)) return { text: "", kind: `skipped: optional ${suffix.slice(1)} parser is not installed` };
    return { text: raw, kind: suffix.slice(1).toUpperCase() || "text" };
  } catch (error) {
    return { text: "", kind: `skipped: ${error.message}` };
  }
}

function buildContext({ output, files = [] }) {
  if (!output) throw new Error("Context output path is required");
  if (!Array.isArray(files) || !files.length) throw new Error("At least one context file is required");
  const sections = ["# Penetration Testing Context", "", "Generated from imported source material.", ""];
  let parsed = 0;
  for (const rawPath of files) {
    const filePath = path.resolve(String(rawPath));
    const result = extract(filePath);
    sections.push(`## ${path.basename(filePath)}`, "", `_Source: \`${filePath}\` · ${result.kind}_`, "");
    if (result.text.trim()) {
      sections.push("```text", result.text.trim().slice(0, MAX_TEXT_CHARS), "```", "");
      parsed += 1;
    } else sections.push(`> ${result.kind}`, "");
  }
  const target = path.resolve(output);
  fs.mkdirSync(path.dirname(target), { recursive: true });
  fs.writeFileSync(target, sections.join("\n"), "utf8");
  return { ok: true, parsed, total: files.length, output: target };
}

module.exports = { extract, buildContext, MAX_FILE_BYTES, MAX_TEXT_CHARS };
